### PR TITLE
Issue #28: Frontend -- Authentication

### LINKED ISSUE #28: Frontend -- Signup and Signin Screens
# Issue Template

## Prerequisites

Please answer the following questions for yourself before submitting an issue.

## Type of PR

- [x] Feature Request
- [ ] Report Bug
- [ ] Documentation Update
- [ ] Other

## Context

Signup and Signin screens for frontend

## Expected Behavior

Component for signup and signin screen should be rendered with implemented UI

## Current Behavior

No UI code is implemented

## Failure Information (for bugs)

Please help provide information about the failure if this is a bug. If it is not a bug, please remove the rest of this template.

## Check

- [x] I have followed issue creation guidelines
- [x] I have attached proper label with the issue
- [x] The issue is assigned to a assignee

### PR DESCRIPTION
**Pull request by basantech89**
# Pull Request Template

## Description

This PR addressed following changes
- [x] Authentication feature
- [x] Use of request library to fetch data from server
- [x] Styling

Resolves #28 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Unit Test
- [ ] Integration Test

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: client/src/Main.js [hunk 1]
```diff
@@ -2,21 +2,21 @@ import React from 'react';
...
-      <Route exact path="/signup" component={SignUp} />
-      <Route exact path="/signin" component={SignIn} />
-      <Route exact path="/create" component={CreateNewRelay} />
-      <Route exact path="/taskhistory" component={TaskHistory} />
-      <Route exact path="/myapps" component={MyApps} />
-      <Route exact path="/taskhistory/:id" component={TaskHistoryDetail} />
+      <Route path="/signin" component={SignIn} />
+      <Route path="/signup" component={SignUp} />
+      <Route path="/create" component={CreateNewRelay} />
+      <Route path="/taskhistory" component={TaskHistory} />
```
Comment (rishabh1403) [inline client/src/Main.js]: may want to use exact here?

File: client/src/components/common/Form.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import React, { createRef } from 'react';
+import PropTypes from 'prop-types';
+import {
+  Form as ReactstrapForm, Input, Col, Button, FormGroup, Label,
```
Comment (rishabh1403) [inline client/src/components/common/Form.js]: put in separate lines please.. the imports

File: client/src/components/common/Form.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  const RenderFields = () => formFields.map((field, idx) => {
+    const {
+      type, label, id, placeholder, color, className,
+    } = field;
+    const ref = createRef();
+    const key = idx.toString();
+    const attributes = {
+      type, name: id, id, placeholder, innerRef: ref, required: 'required',
+    };
+    const ValidatedInput = () => (type === 'password' ? <Input {...attributes} minLength="8" /> : <Input {...attributes} />);
```
Comment (rishabh1403) [inline client/src/components/common/Form.js]: line too long, please add line breaks

File: client/src/components/common/Form.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  };
+
+  const RenderFields = () => formFields.map((field, idx) => {
+    const {
+      type, label, id, placeholder, color, className,
+    } = field;
+    const ref = createRef();
+    const key = idx.toString();
+    const attributes = {
+      type, name: id, id, placeholder, innerRef: ref, required: 'required',
```
Comment (rishabh1403) [inline client/src/components/common/Form.js]: please add line breaks

File: client/src/components/common/index.js [hunk 5]
```diff
@@ -5,3 +5,4 @@ export { default as LabelItem } from './LabelItem';
 export { default as RowCol } from './RowCol';
 export { default as CardWithImage } from './CardWithImage';
 export { default as CustomModal } from './CustomModal';
+export { default as Form } from './Form';
```
Comment (rishabh1403) [inline client/src/components/common/index.js]: `export * as Form from './Form'`

will this work? I may be wrong, but give it a try once :)

File: client/src/utils/formUtils.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+const createInputField = (type, id, label, placeholder, inputType = 'input') => ({
+  type, inputType, id, label, placeholder,
+});
+
+const createButtonField = (text, onClick, className = '', inputType = 'button', color = 'primary') => ({
```
Comment (rishabh1403) [inline client/src/utils/formUtils.js]: pass params as an object when there are so many params

File: client/src/utils/requestEPLib.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+const baseURL = 'http://localhost:3001';
```
Comment (rishabh1403) [inline client/src/utils/requestEPLib.js]: this should be a config/env/constant

File: client/src/utils/requestLib.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import qs from 'qs';
+import request from 'request';
+
+const makeRequest = (options) => {
```
Comment (rishabh1403) [inline client/src/utils/requestLib.js]: consider async / await please

File: client/src/components/common/index.js (hunk 5 above)
Comment (basantech89) [reply client/src/components/common/index.js]: Below error occurres if we do this:-
2019-12-13-1576236460_screenshot_1366x768

File: client/src/components/common/Form.js (hunk 3 above)
Comment (basantech89) [reply client/src/components/common/Form.js]: Refactored this

File: client/src/components/common/Form.js (hunk 2 above)
Comment (basantech89) [reply client/src/components/common/Form.js]: done

File: client/src/utils/requestEPLib.js (hunk 7 above)
Comment (basantech89) [reply client/src/utils/requestEPLib.js]: done

File: client/src/components/Authentication/UserForm.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { NavLink } from 'react-router-dom';
+import { createInputField, createButtonField } from '../../utils/formUtils';
+import { Form } from '../common';
+
+const UserForm = (props) => {
+  const { onClick, action } = props;
+  const emailInput = createInputField('email', 'email', 'Email', 'username@example.com');
```
Comment (rishabh1403) [inline client/src/components/Authentication/UserForm.js]: hardcoded username and password?

Comment (basantech89) [reply client/src/components/Authentication/UserForm.js]: No, that's just a placeholder

**Pull request merged**